File: web/src/app/core/auth/auth.service.ts
import { Injectable } from '@angular/core';
import {
  confirmResetPassword,
  confirmSignUp,
  fetchAuthSession,
  getCurrentUser,
  resendSignUpCode,
  resetPassword,
  signIn,
  signOut,
  signUp,
} from 'aws-amplify/auth';

/**
 * The only place that talks to Cognito.
 *
 * <p>Everything else in the application deals in "signed in or not" and an access token, so
 * swapping the identity provider would not reach past this file.
 */
@Injectable({ providedIn: 'root' })
export class AuthService {
  async signUp(email: string, password: string): Promise<void> {
    await signUp({
      username: email,
      password,
      options: { userAttributes: { email } },
    });
  }

  async confirmSignUp(email: string, code: string): Promise<void> {
    await confirmSignUp({ username: email, confirmationCode: code });
  }

  async resendConfirmationCode(email: string): Promise<void> {
    await resendSignUpCode({ username: email });
  }

  /**
   * Returns false when Cognito needs something more before the session exists — an unconfirmed
   * address, or a challenge. The caller routes on that rather than treating it as an error.
   */
  async signIn(email: string, password: string): Promise<boolean> {
    const result = await signIn({ username: email, password });
    return result.isSignedIn;
  }

  async signOut(): Promise<void> {
    await signOut();
  }

  async startPasswordReset(email: string): Promise<void> {
    await resetPassword({ username: email });
  }

  async confirmPasswordReset(email: string, code: string, newPassword: string): Promise<void> {
    await confirmResetPassword({ username: email, confirmationCode: code, newPassword });
  }

  async isSignedIn(): Promise<boolean> {
    try {
      await getCurrentUser();
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Amplify refreshes the access token here when it is close to expiry, which is why every
   * request asks for it rather than caching one.
   */
  async accessToken(): Promise<string | null> {
    try {
      const session = await fetchAuthSession();
      return session.tokens?.accessToken?.toString() ?? null;
    } catch {
      return null;
    }
  }
}

File: ios/Packages/SimplicityAuth/Sources/SimplicityAuth/AuthService.swift
import Foundation
import Mockable

/// The only place in the app that talks to Cognito.
///
/// Everything else deals in "signed in or not" and an access token, so swapping the identity
/// provider would not reach past this file — the same boundary the web draws in
/// `web/src/app/core/auth/auth.service.ts`.
@Mockable
public protocol AuthService: AnyObject, Sendable {

    func signUp(email: String, password: String) async throws
    func confirmSignUp(email: String, code: String) async throws
    func resendConfirmationCode(email: String) async throws

    /// Returns false when Cognito needs something more before a session exists — an unconfirmed
    /// address, or a challenge. The caller routes on that rather than treating it as an error.
    func signIn(email: String, password: String) async throws -> Bool

    func signOut() async
    func startPasswordReset(email: String) async throws
    func confirmPasswordReset(email: String, code: String, newPassword: String) async throws
    func isSignedIn() async -> Bool

    /// Nil rather than a throw when signed out: the caller is usually a request builder that
    /// should proceed unauthenticated, not fail.
    func accessToken() async -> String?
}
